File: assets/api/add_source.py
import urllib, urllib2, cookielib
import json
import datetime
import random

# this is a sample python client for the JSON API for adding a variable
# author: Jiri Kadlec

username = 'admin'
password = 'password'

sourceID = random.random() * 1000
sourceName = "python test %s" % sourceID

#STEP 1: prepare the uploaded data in JSON format
data = {
    "user": username,
    "password": password,
    "organization": "python source %s" % sourceID,
    "description": "uploaded from python: %s" % sourceID,
    "link": "http://example.com",
    "name": "name %s" % sourceID,
    "phone": "012-345-6789",
    "email": "test@gmail.com",
    "address": sourceName,
    "city": sourceName,
    "state": sourceName,
    "zipcode": sourceName,
    "citation": "uploaded from python as a test",
    "metadata": 10
}
postdata = json.dumps(data)


#STEP 2: post the data values to HydroServer and check response status
uploadURL = 'http://worldwater.byu.edu/app/index.php/default/services/api/sources'
req = urllib2.Request(uploadURL)
req.add_header('Content-Type', 'application/json')

try:
    response = urllib2.urlopen(req, postdata)
    print response.read()

except urllib2.HTTPError, e:
    print e.code
    print e.msg
    print e.headers
    print e.fp.read()

File: assets/api/AddSource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using System.Net;
using System.Windows.Forms;

namespace APITest
{
    class SourceHelper
    {
        private void AddSource()
        {
            string url = "http://worldwater.byu.edu/app/index.php/default/services/api/sources";

            System.Net.HttpWebRequest request = (HttpWebRequest)System.Net.HttpWebRequest.Create(url);
            request.Method = "POST";

            request.ContentType = "application/json";

            using (var streamWriter = new StreamWriter(request.GetRequestStream()))
            {
                string json = @"{'user': 'admin',
                        'password': 'password',
                        'organization':'test004',
                        'description':'test',
                        'link':'link001',
                        'name':'teva',
                        'phone':'123-456-7891',
                        'email':'email@email.com',
                        'address':'address001',
                        'city':'city001',
                        'state':'Utah',
                        'zipcode':'zipcode001',
                        'citation':'citation001',
                        'metadata':10 }";

                json = json.Replace("'", "\"");
                streamWriter.Write(json);
                streamWriter.Flush();
                streamWriter.Close();
            }


            try
            {
                using (WebResponse response = request.GetResponse())
                {
                    using (var streamReader = new StreamReader(response.GetResponseStream()))
                    {
                        var result = streamReader.ReadToEnd();
                        MessageBox.Show(result);
                    }
                }
            }
            catch (WebException ex)
            {
                using (WebResponse response = ex.Response)
                {
                    HttpWebResponse httpResponse = (HttpWebResponse)response;
                    string errorCode = string.Format("Error code: {0} ", httpResponse.StatusCode);
                    using (Stream data = response.GetResponseStream())
                    using (var reader = new StreamReader(data))
                    {
                        string text = reader.ReadToEnd();
                        MessageBox.Show(errorCode + text);
                    }
                }
            }
        }
    }
}
